### ISSUE TITLE
[DevShares 0.5.0] 10 assert_exception when trying to wallet_transfer from TITAN account

### ISSUE THREAD

**Issue by robrigo** [BitShares/bitshares#1229]
Hey team. Seeing some weirdness so I figured I'd report it.

I imported my private key for [account_A], an account that I created via the DevShares 0.5.0 GUI. wallet_dump_private_key [account_A] returns a key. Could wallet_transfer broken in this release or maybe my key is just messed up?

```
[USER] (unlocked) >>> wallet_transfer 497 DVS [account_A] [account_A]
10 assert_exception: Assert Exception
key->has_private_key():
    {}
    th_a  wallet.cpp:1572 get_private_key

    {"addr":"[pub_key_1]"}
    th_a  wallet.cpp:1574 get_private_key

    {"recipient":{"id":347,"name":"[account]","public_data":{"version":"v0.5.0"},"owner_key":"[pub_key_2]","active_key_history":[["2015-01-09T19:18:00","[pub_key_3]"]],"registration_date":"2015-01-09T19:18:00","last_update":"2015-01-09T22:49:00","delegate_info":{"votes_for":0,"pay_rate":75,"signing_key_history":[[25750,"[pub_key_3]"]],"last_block_num_produced":0,"next_secret_hash":null,"pay_balance":0,"total_paid":0,"total_burned":0,"blocks_produced":0,"blocks_missed":0},"meta_data":null},"amount":{"amount":49700000,"asset_id":0},"memo":""}
    th_a  transaction_builder.cpp:229 deposit_asset

    {}
    th_a  common_api_client.cpp:4008 wallet_transfer

    {"command":"wallet_transfer"}
    th_a  cli.cpp:629 execute_command
```

Using wallet_transfer_to_public_account worked sending TO and FROM the same account but the receiver is UNKNOWN in the transaction history:

```
[USER] (unlocked) >>> wallet_account_transaction_history [account_A]
TIMESTAMP           BLOCK     FROM                TO                  AMOUNT                  MEMO                                        BALANCE                 FEE                 ID
==============================================================================================================================================================================================
...
2015-01-12T03:38:34 45788     [account_A]    UNKNOWN             497.00000 DVS           To: [pub_key_1]                             0.00000 DVS             1.00000 DVS         427c7f6f
```

I regenerated 5000 keys for [account_A] and got the same results.

**Comment by vikramrajkumar** [comment]
Possibly related: #1220

**Comment by theoreticalbts** [comment]
I see the account in question is a delegate account with different active and owner keys.

This leads me to suspect you are keeping the owner key in cold storage and issuing this command from an account that only has access to the active key.

However, the key needed to withdraw a balance is set *at the time the coins are sent*.  The consequence of that is as follows:

- If you change your active key, you must then transfer your entire balance to yourself *from a wallet with access to the old active key*, otherwise any parts of your balance addressed to your old active key will be inaccessible to a wallet which only has the new active key.

In other words, at the blockchain level, balances are owned by keys, not accounts.  Changing your account's active key doesn't change the key that owns existing balances, that requires a separate transaction!

**Comment by theoreticalbts** [comment]
I put together a test which reproduces this problem.

Apparently if you create a wallet, set active key different from owner key, open a new wallet, import only the active key, and transact your entire balance to yourself, it throws this assertion.

And that's *definitely* a bug.

**Comment by theoreticalbts** [comment]
OP was just the tip of the iceberg.  It's made me realize that active/owner keys present multiple issues that stretch from client architecture to UI.

The core problem is, as I stated earlier, at the blockchain level coins are always sent to addresses, not accounts.  As a consequence, in the most general case:

- An account has had an owner key and N distinct active keys (not counting the owner key)
- Any subset of those N+1 keys may exist in the wallet
- Any subset of those N+1 keys may control balances on the blockchain

The two subsets above are not necessarily the same!

So when using an asset for any purpose -- transfer, market operations, fees, whatever -- the client must look at balances available to *all keys it controls which have ever been active / owner keys for the account*, and in the case of a TITAN account, all funds available to TITAN-derived addresses as well.

We need a way to represent in the UI the following conditions:

- The wallet controls private keys that have balances, but are not the most recent active key.  In this case, the user may wish to consolidate those balances on either the most recent active key or the owner key.

- Some of the N+1 keys control balances on the blockchain, but the wallet does not have the corresponding private keys.  The user may wish to know about those balances, e.g. to keep track of how much is in their cold storage, but they must be displayed separately from the balances the wallet actually controls.  This mostly applies to public accounts:  While any of the N+1 keys can also have a balance in the case of a TITAN account, the wallet has no way of even knowing those balances belong to the user's account unless it has the corresponding private keys.

But to start with, we need to at least make sure the CLI wallet properly handles all cases.  In particular, all transactions which use an account's balances in any way, must consider balances to be available for use in transactions *if and only if*:

- They belong to an address representing a pubkey that's ever been an active / owner key for an account (or a TITAN-derived pubkey),
- AND the wallet controls the private key belonging to that address.

This is much less simple as semi-mechanically substituting `active_key` for `owner_key` -- the client code for *every possible transaction type* must be aware of this issue.

**Comment by robrigo** [comment]
Interesting analysis of the problem theoretical. 

I noticed that I am able to subvert this assert_exception by using wallet_transfer_to_public_account with the same parameters above. The trx happens and balances are sent appropriately.

I believe that I deleted the owner key for this delegate! I was mucking around with the DVS GUI client and was having issues syncing and ended up just deleting the app data folder to force a resync. Afterward I imported the account private keys back into a newly created wallet. I should have created a copy of the directory even though its just the test net, but oh well. I can still move my funds using wallet_transfer_to_public_account as stated above, but is that going to be a problem going forward for this account?

**Comment by theoreticalbts** [comment]
@vikramrajkumar assures me the architecture does gather balances from all private keys the wallet has access to.  The problem seems to be in `wallet_transfer_from` attempting to encrypt the memo with the owner key, regardless of whether the owner key exists in the wallet.

I think this behavior should be replaced as follows:

- Use the account owner key if it exists in the wallet.
- Otherwise, sign the memo with the most recent account active key that exists in the wallet.
